### PR TITLE
RFC: kubeadm: build dag executor

### PR DESCRIPTION
**Pull request by mikedanese**
This PR adds a simple dag executor capable of executing dependent tasks in parallel without violating dependencies. I want to evaluate if this would be a useful abstraction to model kubeadm task execution. In this model, each task that kubeadm performs is a `Unit` with a name, dependencies, and an action. Units can compose easily and naturally.

Specifically I want to evaluate if this model can improve:
* testability: unit tests!
* performance through concurrency
* code readability/maintainability: phases are one or more units that expose a single logical unit.
* log consistency

@kubernetes/sig-cluster-lifecycle-misc

### CODE REVIEW

**Comment by timothysc** [comment]
So long as we don't goto plaid here.  I think we need to scope bound the phases and execution order.  The last thing I want is to reinvent is too many pieces of CM-tools.

**Review by luxas — COMMENTED** [review]
I like the idea of having the general `Unit` interface for every phase, that is kind of the end goal for the phases refactoring.

But I'm not sure we really need or should have the concurrency. Seems like the key outcome of that would be race conditions. I think something like this would be as useful:
```go
units := []Unit{certificates, kubeconfig, etcd, controlplane,...}
for _, unit := units {
     unit.Execute(context, os.Stdout)
}
```

WDYT?

**Comment by mikedanese** [comment]
I don't like that for a few reasons:

* We should strive to improve performance of cluster deployment.
* We should not use single thread of execution as a safety blanket to not fully understand dependencies between phases.

I'm in favor of explicitly declaring dependencies between units. We still need to dependency sort in that case. We can roll this out single threaded and phase in concurrency later on though.

**Comment by luxas** [comment]
> We can roll this out single threaded and phase in concurrency later on though.

I guess I'm fine with that then.

As long as we don't grow into systemd-like functionality, I'm ok with this ;)

> understand dependencies between phases.

This is something I'm working on as much as I have time for it. Definitely want to make even more progress on it this cycle and could indeed use help

**Comment by mikedanese** [comment]
Even with a single thread of execution, there's still value in modelling kubeadm execution as a DAG.

File: cmd/kubeadm/app/util/executor/executor.go [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+			}
+			if len(n.in) == 0 {
+				roots = append(roots, n.name)
+			}
+		}
+		roots = roots[1:]
+	}
+
+	if len(sorted) != len(nodes) {
+		return nil, fmt.Errorf("circular dependency")
```
Comment (mikedanese) [inline cmd/kubeadm/app/util/executor/executor.go]: This error would be more informative if we used a DFS instead of Kahn's.

**Pull request closed without merging**